### PR TITLE
feat: Added new find-stemcell-version command and modified update-stemcell command

### PR DESCRIPTION
**Pull request by ram-pivot**

### CODE REVIEW

File: commands/find_stemcell_version.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	var productSlug=""
+
+	//Get Stemcell OS and major from Kilnfile
+	if kilnfile.Stemcell.OS == "ubuntu-xenial" {
+		productSlug = "stemcells-ubuntu-xenial"
+	} else if kilnfile.Stemcell.OS == "windows2019" {
+		productSlug = "stemcells-windows-server"
+	}
```
Comment (crhntr) [inline commands/find_stemcell_version.go:60]: Since we already have the idea of a product slug in the Kilnfile. I think it might be easy enough to edit the Kilnfile stemcell section to specify the slug in addition to the name and version constraint.

File: commands/find_stemcell_version.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    if productSlug == "" {
+		return fmt.Errorf(ErrStemcellOSInfoMustBeValid)
+	} else if kilnfile.Stemcell.Version == "" {
```
Comment (crhntr) [inline commands/find_stemcell_version.go:64]: I find it easier to follow code when the happy path is not indented. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: commands/find_stemcell_version_test.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+					Expect(executeErr).NotTo(HaveOccurred())
+					Expect((&writer).String()).To(ContainSubstring("\"456.118\""))
+					Expect((&writer).String()).To(ContainSubstring("\"remote_path\":\"network.pivotal.io\""))
+					Expect((&writer).String()).To(ContainSubstring("\"source\":\"Tanzunet\""))
+				})
+			})
+		})
+	})
+
+	Describe("Execute", func() {
```
Comment (crhntr) [inline commands/find_stemcell_version_test.go:159]: This should probably be named "ExtractMajorVersion" the previous `Describe` is testing "Execute".

File: commands/update_stemcell.go [hunk 4]
```diff
@@ -3,21 +3,21 @@ package commands
...
 	"log"
+	"strings"
 
 	"github.com/pivotal-cf/jhanda"
-	"github.com/pivotal-cf/kiln/builder"
 	"github.com/pivotal-cf/kiln/fetcher"
-	"github.com/pivotal-cf/kiln/helper"
 	"github.com/pivotal-cf/kiln/release"
+	"github.com/Masterminds/semver"
 	"gopkg.in/src-d/go-billy.v4/osfs"
```
Comment (crhntr) [inline commands/update_stemcell.go:12]: (feel free to ignore this comment)

I usually sort my import statements into these ordered blocks (based on [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) but extended a little bit)
- standard library
- dot imports (https://github.com/golang/go/wiki/CodeReviewComments#import-dot)
- test helper/matcher packages
- external libraries
- packages in the current module

File: fetcher/pivnet.go [hunk 5]
```diff
@@ -27,58 +28,68 @@ type Pivnet struct {
...
+	if majorStemcellVersion == "" {
+		return "", ErrStemcellMajorVersionMustNotBeEmpty
 	}
+
 	locator := url.URL{
 		Scheme: "https",
 		Host:   "network.pivotal.io",
-		Path:   path.Join("/api/v2/products", string(slug), "releases"),
+		Path:   fmt.Sprintf("/api/v2/products/%s/releases/latest?major=%s", slug, majorStemcellVersion),
+	}
```
Comment (crhntr) [inline fetcher/pivnet.go:52]: 💯   Nice! I like using `url.URL`. It is uncommon but I feel it is easier to follow than a really long `fmt.Sprintf`.

The RawQuery is a bit oddly named but it would be where you put the query for this URL instead of appending to Path.

File: commands/find_stemcell_version.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		}
+
+		//Get stemcell version from pivnet
+		stemcellVersion, err := cmd.pivnetService.StemcellVersion(productSlug, majorVersion)
+
+		if err != nil {
+			return err
+		}
+
+		stemcellVersionJson, _ := json.Marshal(stemcellVersionOutput{
```
Comment (crhntr) [inline commands/find_stemcell_version.go:81]: I would at least return the marshal error. It is rare that something would go wrong but ignoring errors should be a rare edge case. I would not write a test for such an error check.

File: commands/find_stemcell_version.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+			Version: stemcellVersion,
+			Source: "Tanzunet",
+			RemotePath: TanzunetRemotePath,
+		})
+		cmd.outLogger.Println(string(stemcellVersionJson))
+	}
+	return nil
+}
+
+func ExtractMajorVersion(version string) (string, error) {
```
Comment (crhntr) [inline commands/find_stemcell_version.go:91]: (this is a really loosly held opinion... feel free to ignore it)

Version constraints from this package may not match what the code below seems to expect. It might make sense to fetch the versions from Pivnet and then see which version matches the constraint. Another option would be to instead of specifying the version have the key specify the version line. For example:

```go
package cargo

type StemcellConstraint struct {
	Slug         string `yaml:"slug"`
	MajorVersion int    `yaml:"version_line"`
}
```

**Review by crhntr — CHANGES_REQUESTED** [review]
Looks solid for the most part 👍 

I think it might be good to specify the slug in the Kilnfile's stemcell_critera section instead of having control flow in the command.  For example you can extend the type like this:

```go
package cargo

//...

type PublishedStemcell struct {
	Stemcell
	ProductSlug string `yaml:"slug"`
}
```
Then use this variant in the Kilnfile struct.

In addition to the review comments. These are a few small things I noticed:
- The Go code could use a `go fmt ./...`.
- My linter is warning about error messages starting with an upper case. I kind of like how they are currently and would not mind merging them with upper case letters. https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: commands/find_stemcell_version.go (hunk 1 above)
Comment (ram-pivot) [reply commands/find_stemcell_version.go:62]: Yhea I keep forgetting the go fmt ./... thanks for noticing that. Also as discussed I am not making the product slug changes for now.

File: commands/find_stemcell_version.go (hunk 2 above)
Comment (ram-pivot) [reply commands/find_stemcell_version.go]: Thanks for the link. Yhea I agree I changed this one.

File: commands/find_stemcell_version_test.go (hunk 3 above)
Comment (ram-pivot) [reply commands/find_stemcell_version_test.go]: Yhea this one was a typo 👍

File: commands/update_stemcell.go (hunk 4 above)
Comment (ram-pivot) [reply commands/update_stemcell.go]: Yhea I liked this. Usually I run tools for other languages that does this for me. Probably I need to setup something for golang.

For now I manually updated the imports.

File: fetcher/pivnet.go (hunk 5 above)
Comment (ram-pivot) [reply fetcher/pivnet.go:54]: That is cool to know. I added the RawQuery.  I ran into issue with the query parameter adding additional / to the value. So have to use fmt.Sprintf.

File: commands/find_stemcell_version.go (hunk 6 above)
Comment (ram-pivot) [reply commands/find_stemcell_version.go]: This one makes sense. I am returning the error now.

File: commands/find_stemcell_version.go (hunk 7 above)
Comment (ram-pivot) [reply commands/find_stemcell_version.go:100]: Yhea for now the Kilnfile has options to pin a version. May be we can later expand if we want robust semver rules. I assumed for Stemcell we might not have very complex rules and just limited based on the major alone. But good point where we might have to consider for complex version constraints.

**Comment by ram-pivot** [comment]
@crhntr When you get a chance please check and let me know if anything else needs to be updated. Thanks again for taking the time to review the PR.

**Review by crhntr — APPROVED** [review]
